torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -697,13 +721,27 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
         inputs[i].storage().data_ptr(), ncclStream);
   }
 
+  std::unique_lock<std::mutex> lock(checkFutObjMutex_);
+  // Create and store a CheckFutureWork object to be used in
+  // ncclKernelCompletionCallback function.
+  auto checkFutObj = std::make_shared<CheckFutureWork>(
+      work, outputs, checkFutObjs_, checkFutObjMutex_);
+  checkFutObjs_->insert(checkFutObj);
+
   {
     AutoNcclGroup nccl_group_guard;
     for (size_t i = 0; i < inputs.size(); ++i) {
       gpuGuard.set_index(devices[i].index());
       at::cuda::CUDAStream& ncclStream = ncclStreams_[key][i];
       C10D_NCCL_CHECK(
           fn(inputs[i], outputs[i], ncclComms[i]->getNcclComm(), ncclStream));
+
+      // Add a cudaStreamCallback on the same stream. CUDA will automatically
+      // call ncclKernelCompletionCallback once the appropriate NCCL kernel
+      // is done and workNCCL's Future work will be marked as completed
+      // using this callback.
+      cudaStreamAddCallback(
```

Comment:
I think the outputs not being populated issue might come from this: https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/groups.html
"Caution: When called inside a group, stream operations (like ncclAllReduce) can return without having enqueued the operation on the stream. Stream operations like cudaStreamSynchronize can therefore be called only after ncclGroupEnd returns."

Maybe we need to move this cudaStreamAddCallback calls outside of the `nccl_group_guard`?

Reply:
I'm trying that. Thanks,

Reply:
The [docs] for `cudaStreamAddCallback` say "Adds a callback to be called on the host after all currently enqueued items in the stream have completed", so my reasoning in conjunction with the above link was that if the operation is not enqueued, then there is a chance, according to the documentation, that the callback would fire, because the condition for the callbacks being called is (vacuously) true. Although, I may not be interpreting the docs correctly and maybe this can't happen in practice.